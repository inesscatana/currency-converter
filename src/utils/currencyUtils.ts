/**
 * Extracts the target currency code from a currency pair.
 * Example: "AAVE-USD" => "AAVE", "$ADS-USD" => "ADS"
 *
 * @param pair - The currency pair string (e.g., "BTC-USD").
 * @param base - The base currency to be removed (e.g., "USD").
 * @returns The extracted currency code without the base currency.
 */
export function getCurrencyCodeFromPair(pair: string, base: string): string {
  return pair.replace(`-${base}`, '').replace(/^\$/, ''); // Removes "-USD" and "$" prefix
}

/**
 * Calculates the converted amount based on the exchange rate.
 * Example: (1.25, 100) => "125.0000"
 *
 * @param currencyValue - The exchange rate of the target currency.
 * @param inputAmount - The amount to be converted.
 * @returns The converted value formatted to 4 decimal places.
 */
export function getCalculatedRate(
  currencyValue: number,
  inputAmount: number
): string {
  return (currencyValue * inputAmount).toFixed(4);
}

/**
 * Formats a numeric string into a localized number with thousands separators.
 *
 * - Removes any non-numeric characters except for one decimal point.
 * - Ensures proper formatting with thousands separators and a defined number of decimal places.
 * - Returns an empty string if the input is invalid or empty.
 *
 * Example:
 * ```ts
 * formatToLocaleString("1234567.89") // "1,234,567.89"
 * formatToLocaleString("1000000.123456", 4) // "1,000,000.1235"
 * ```
 *
 * @param entry - The input string containing a numeric value.
 * @param decimalPlaces - The number of decimal places to retain (default: 6).
 * @returns The formatted number as a string, or an empty string if invalid.
 */
export function formatToLocaleString(entry: string, decimalPlaces = 6): string {
  if (!entry) return '';

  let sanitizedValue = entry.replace(/[^0-9.]/g, ''); // Remove non-numeric characters except '.'
  sanitizedValue = sanitizedValue.replace(/\.(?=.*\.)/g, ''); // Ensure only one decimal point

  const number = parseFloat(sanitizedValue);
  return isNaN(number)
    ? ''
    : number.toLocaleString('en-US', { maximumFractionDigits: decimalPlaces });
}
