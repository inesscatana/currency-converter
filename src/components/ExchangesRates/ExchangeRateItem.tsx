import { useMemo } from 'react';
import { getCalculatedRate } from '../../utils/currencyUtils';

interface ExchangeRateItemProps {
  currency: string;
  value: number;
  amount: string;
  getCurrencyIcon: (currency: string) => string;
}

const ExchangeRateItem = ({
  currency,
  value,
  amount,
  getCurrencyIcon,
}: ExchangeRateItemProps) => {
  const calculatedRate = useMemo(
    () => getCalculatedRate(value, parseFloat(amount)),
    [value, amount]
  );

  return (
    <li className="flex justify-between items-center px-6 sm:px-4">
      {/* Amount */}
      <span className="text-gray-900 font-semibold text-2xl leading-7 tracking-tight min-w-[7rem] sm:min-w-[6rem] text-left">
        {calculatedRate}
      </span>

      {/* Currency Icon + Label */}
      <div className="flex items-center gap-2 px-2 py-2 min-w-[7.5rem] sm:min-w-[6.5rem] justify-center">
        <img
          src={getCurrencyIcon(currency)}
          alt={currency}
          className="w-6 h-6 sm:w-5 sm:h-5 rounded-full"
        />
        <span className="text-gray-700 font-semibold text-lg sm:text-sm leading-6">
          {currency}
        </span>
      </div>
    </li>
  );
};

export default ExchangeRateItem;
