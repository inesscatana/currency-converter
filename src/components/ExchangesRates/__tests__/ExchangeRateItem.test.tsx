import { render, screen } from '@testing-library/react';
import ExchangeRateItem from '../ExchangeRateItem';
import { getCalculatedRate } from '../../../utils/currencyUtils';

jest.mock('../../../utils/currencyUtils', () => ({
  getCalculatedRate: jest.fn((value, amount) => (value * amount).toFixed(4)),
}));

describe('ExchangeRateItem Component', () => {
  const mockGetCurrencyIcon = jest.fn((currency) => `icon-${currency}.png`);

  const props = {
    currency: 'USD',
    value: 1.25,
    amount: '100',
    getCurrencyIcon: mockGetCurrencyIcon,
  };

  beforeEach(() => {
    jest.clearAllMocks(); // âœ… Reset mocks before each test
  });

  it('renders correctly with given props', () => {
    render(<ExchangeRateItem {...props} />);

    // Check if the calculated rate is displayed correctly
    const calculatedRate = getCalculatedRate(
      props.value,
      parseFloat(props.amount)
    );
    expect(screen.getByText(calculatedRate)).toBeInTheDocument();

    // Check if currency label is displayed
    expect(screen.getByText(props.currency)).toBeInTheDocument();
  });

  it('renders the correct currency icon', () => {
    render(<ExchangeRateItem {...props} />);

    const currencyIcon = screen.getByRole('img', { name: props.currency });
    expect(currencyIcon).toHaveAttribute(
      'src',
      mockGetCurrencyIcon(props.currency)
    );
  });

  it('ensures the calculation function is called correctly', () => {
    render(<ExchangeRateItem {...props} />);

    expect(getCalculatedRate).toHaveBeenCalledWith(
      props.value,
      parseFloat(props.amount)
    );
  });

  it('matches the snapshot', () => {
    const { container } = render(<ExchangeRateItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('only recalculates when props change', () => {
    const { rerender } = render(<ExchangeRateItem {...props} />);

    rerender(<ExchangeRateItem {...props} />);
    rerender(<ExchangeRateItem {...props} />);

    expect(getCalculatedRate).toHaveBeenCalledTimes(1);
  });
});
