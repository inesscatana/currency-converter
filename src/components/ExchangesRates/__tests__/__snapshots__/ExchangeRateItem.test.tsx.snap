// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeRateItem Component matches the snapshot 1`] = `
<div>
  <li
    class="flex justify-between items-center px-6 sm:px-4"
  >
    <span
      class="text-gray-900 font-semibold text-2xl leading-7 tracking-tight min-w-[7rem] sm:min-w-[6rem] text-left"
    >
      125.0000
    </span>
    <div
      class="flex items-center gap-2 px-2 py-2 min-w-[7.5rem] sm:min-w-[6.5rem] justify-center"
    >
      <img
        alt="USD"
        class="w-6 h-6 sm:w-5 sm:h-5 rounded-full"
        src="icon-USD.png"
      />
      <span
        class="text-gray-700 font-semibold text-lg sm:text-sm leading-6"
      >
        USD
      </span>
    </div>
  </li>
</div>
`;
